solidity/contracts/liquidity-protection/LiquidityProtection.sol

```diff
@@ -1181,12 +1229,17 @@ contract LiquidityProtection is TokenHandler, ContractRegistryClient, Reentrancy
         internal view returns (uint256)
     {
         Fraction memory poolRate = poolTokenRate(_poolToken, _reserveToken);
-        Fraction memory poolFactor = poolTokensFactor(_addRate, _removeRate);
+        uint256 n = Math. ceilSqrt(_addRate.d.mul(_removeRate.n)).mul(poolRate.n);
+        uint256 d = Math.floorSqrt(_addRate.n.mul(_removeRate.d)).mul(poolRate.d);
 
-        (uint256 poolRateN, uint256 poolRateD) = Math.reducedRatio(_poolAmount.mul(poolRate.n), poolRate.d, MAX_UINT128);
-        (uint256 poolFactorN, uint256 poolFactorD) = Math.reducedRatio(poolFactor.n, poolFactor.d, MAX_UINT128);
+        uint256 x = n * _poolAmount;
+        if (x / n == _poolAmount) {
+            return x / d;
```

Comment:
`d` can't be 0 (because of rounding for example)?

Reply:
No.
Functions `floorSqrt` and `ceilSqrt` are implemented for positive input only (and documented as such).
For the smallest legal input (1), the output is larger than 0 (also 1).
So a divide-by-zero revert (invalid opcode) can take place here only if the input (one of the rates) is 0, which we know is not the case (and if it was the case, then the transaction would revert at a much earlier point).